File: crates/interpreter/src/constants/actions.rs
// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use serde::{Deserialize, Serialize};
use strum_macros::{Display, EnumString, EnumVariantNames, IntoStaticStr};

#[derive(
    Debug,
    Clone,
    Serialize,
    Deserialize,
    PartialEq,
    EnumVariantNames,
    Display,
    EnumString,
    IntoStaticStr,
)]
pub enum InterpretationActionType {
    #[strum(serialize = "NATIVE_RECEIVE")]
    NativeReceive,
    #[strum(serialize = "NATIVE_SEND")]
    NativeSend,
    #[strum(serialize = "ERC20_APPROVE")]
    ERC20Approve,
    #[strum(serialize = "ERC20_RECEIVE")]
    ERC20Receive,
    #[strum(serialize = "ERC20_SEND")]
    ERC20Send,
    #[strum(serialize = "ERC20_MINT")]
    ERC20Mint,
    #[strum(serialize = "ERC20_BURN")]
    ERC20Burn,
    #[strum(serialize = "ERC721_APPROVE")]
    ERC721Approve,
    #[strum(serialize = "ERC721_RECEIVE")]
    ERC721Receive,
    #[strum(serialize = "ERC721_SEND")]
    ERC721Send,
    #[strum(serialize = "ERC721_MINT")]
    ERC721Mint,
    #[strum(serialize = "ERC721_BURN")]
    ERC721Burn,
    #[strum(serialize = "ERC1155_APPROVE")]
    ERC1155Approve,
    #[strum(serialize = "ERC1155_RECEIVE")]
    ERC1155Receive,
    #[strum(serialize = "ERC1155_SEND")]
    ERC1155Send,
    #[strum(serialize = "ERC1155_MINT")]
    ERC1155Mint,
    #[strum(serialize = "ERC1155_BURN")]
    ERC1155Burn,
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::str::FromStr;

    #[test]
    fn test_to_string() {
        let action = InterpretationActionType::NativeReceive;
        let serialized_action = action.to_string();
        assert_eq!(serialized_action, "NATIVE_RECEIVE");
    }

    #[test]
    fn test_from_str() {
        let action_str = "NATIVE_RECEIVE";
        let deserialized_action: Result<InterpretationActionType, _> =
            InterpretationActionType::from_str(action_str);
        assert!(deserialized_action.is_ok());
        assert_eq!(deserialized_action.unwrap(), InterpretationActionType::NativeReceive);
    }

    #[test]
    fn test_round_trip() {
        let original = InterpretationActionType::NativeReceive;
        let serialized = original.to_string();
        let deserialized: Result<InterpretationActionType, _> =
            InterpretationActionType::from_str(&serialized);
        assert!(deserialized.is_ok());
        assert_eq!(original, deserialized.unwrap());
    }
}

File: packages/const/src/actions.ts
// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export enum Action {
  UNKNOWN = 0,
  NATIVE_RECEIVE = 1,
  NATIVE_SEND = 2,
  ERC20_APPROVE = 3,
  ERC20_RECEIVE = 4,
  ERC20_SEND = 5,
  ERC20_MINT = 6,
  ERC20_BURN = 7,
  ERC721_APPROVE = 8,
  ERC721_RECEIVE = 9,
  ERC721_SEND = 10,
  ERC721_MINT = 11,
  ERC721_BURN = 12,
  ERC1155_APPROVE = 13,
  ERC1155_RECEIVE = 14,
  ERC1155_SEND = 15,
  ERC1155_MINT = 16,
  ERC1155_BURN = 17,
}

export const ACTION_LABELS: Record<Action, string> = {
  [Action.UNKNOWN]: "Unknown",
  [Action.NATIVE_RECEIVE]: "Receive",
  [Action.NATIVE_SEND]: "Send",
  [Action.ERC20_APPROVE]: "Approve",
  [Action.ERC20_RECEIVE]: "Receive",
  [Action.ERC20_MINT]: "Mint",
  [Action.ERC20_SEND]: "Send",
  [Action.ERC20_BURN]: "Burn",
  [Action.ERC721_APPROVE]: "Approve",
  [Action.ERC721_RECEIVE]: "Receive",
  [Action.ERC721_SEND]: "Send",
  [Action.ERC721_MINT]: "Mint",
  [Action.ERC721_BURN]: "Burn",
  [Action.ERC1155_APPROVE]: "Approve",
  [Action.ERC1155_RECEIVE]: "Receive",
  [Action.ERC1155_SEND]: "Send",
  [Action.ERC1155_MINT]: "Mint",
  [Action.ERC1155_BURN]: "Burn",
};
